moving_modules.md

```diff
@@ -0,0 +1,315 @@
+there can only be one version of a module installed per-environment (unless you
+use tricks like adding multiple directories to the `modulepath`). To avoid
+naming collisions, I'm proposing we append `_core` to all of the module names
+that we extract, e.g. `augeas_core`. This only changes the name of the module,
+not the names of the types contained within. Also it doesn't eliminate a
+collision, just makes it less likely.
+
+## Builtin Types
+
+The loaders register builtin types based on
+`Puppet::Pops::Loader::StaticLoader#BUILTIN_TYPE_NAMES`. This is an
+optimization since we know the types exist in puppet and we don't need to scan
+the filesystem based on the per-environment modulepath. Any type removed from
+puppet, should be removed from `BUILTIN_TYPE_NAMES`. This will have an impact
+on compiler performance where the timeout is not unlimited.
+
+## Environment Isolation
+
+All of the types extracted from puppet and removed from `BUILTIN_TYPE_NAMES`
+above, will be subject to [environment
+isolation](https://puppet.com/docs/puppet/5.4/environment_isolation.html#environment-isolation)
+issues. Users that install newer versions of modules containing types that are
+also vendored, should use `puppet generate types` to ensure types in one
+environment don't affect other environments. This isn't really a new concern,
+it's just that users haven't had to `puppet generate types` for builtin types
+before.
+
+## Module Dependencies
+
+The autoloader does not rely on module metadata to load types. So any module
+that relies on an type, should just work in Puppet 6 provided the extracted
+module is vendored into puppet-agent (eg `augeas`), or the user installs the module
+(eg `nagios`).
+
+## Module Precedence
+
+### Catalog Compilation
+
+It should be possible to install a new version of a module and have that take
+precedence over the `vendor_modules` directory during compilation (or `puppet
+apply`). That will require a change to the autoloader to search for types.
+
+### Pluginsync
+
+All modules visible in the current environment's modulepath should have their
+lib directories copied to the agent during pluginsync. If there are multiple
+versions of the same type installed (one in the `vendor_modules` directory,
+another in `$codedir/environments/production/modules`), then it's important that
+we pluginsync the same type/provider version as was used during compilation. So
+pluginsync and the compiler need to use the same precedence order when resolving
+types.
+
+### Catalog Application
+
+Agents may have different versions of vendored modules than the server used to
+compile the catalog. The agent should always use the pluginsync'ed version of
+the type and providers instead of whatever version is present in the agent's
+puppet-agent package.
+
+## Agent Compatibility
+
+Pre-6.0 agents will always prefer the builtin version over what was
+pluginsynced. If a new property/parameter is added to the type on the master and
```

Comment:
Unfortunately, pre-6.0 agents currently prefer the pluginsync'ed version over the built-in one.